README.md
# maybe_slow_future
[![pub package](https://img.shields.io/pub/v/maybe_slow_future.svg)](https://pub.dartlang.org/packages/maybe_slow_future)
[![codecov](https://codecov.io/gh/6220119/maybe_slow_future/branch/main/graph/badge.svg)](https://codecov.io/gh/6220119/maybe_slow_future)
[![Build Status](https://travis-ci.org/6220119/maybe_slow_future.svg?branch=main)](https://travis-ci.org/6220119/maybe_slow_future)

Add a callback to tell you the future is taking longer than an expected duration to complete.

## Usage
To use this plugin, add `maybe_slow_future` as a dependency in your pubspec.yaml file.
```yaml
dependencies:
  maybe_slow_future: ^1.0.2
```

### Example

``` dart
/// Remember to import the package to be able to use the [Future.onSlow] extension method
import 'package:maybe_slow_future/maybe_slow_future.dart';

void main() {
  print('Fetching data...');

  const timeToComplete = Duration(seconds: 5);

  /// simulate a long running operation
  final fetchDataFuture = Future.delayed(timeToComplete, () {
    print('Got data after ${timeToComplete}: Hello World!');
  });

  /// use the extension method [Future.onSlow]
  fetchDataFuture.onSlow(Duration(seconds: 2), () {
    print('Be patient! It might take a while to complete...');
  });

  /// or use the [maybeSlowFuture] function
  maybeSlowFuture(
    actualFuture: fetchDataFuture,
    threshold: Duration(seconds: 3, milliseconds: 500),
    onSlowCallback: () {
      print('This is taking longer than expected...');
    },
  );

  /// output:
  /// Fetching data...
  /// Be patient! It might take awhile to complete...
  /// This is taking longer than expected...
  /// Got data after 0:00:05.000000: Hello World!
}
```


CHANGELOG.md
## [1.0.2] - Version bump

* Minor update to document

## [1.0.1] - Version bump

* Minor update to document

## [1.0.0+1] - Initial stable release

* Update document link of the pubspec

## [1.0.0] - Initial stable release

* Implement extension method `Future.onSlow`
* Implement `maybeSlowFuture`


scripts/unit_test.sh
# verified on Dart verson 2.10.1

# assuming standing at the root folder

# generate json coverage data to `coverage` folder
dart test --coverage=coverage

# the next command requires running `dart pub global activate coverage` once
dart pub global run coverage:format_coverage \
  --in coverage/test/maybe_slow_future_test.dart.vm.json \
  --out coverage/lcov.info \
  --lcov \
  --packages=.dart_tool/package_config.json \
  --report-on=lib/

# the next command requires `brew install lcov` on macOS
genhtml -o coverage/html coverage/lcov.info

# launch browser (macOS)
open coverage/html/index.html


test/maybe_slow_future_test.dart
import 'package:fake_async/fake_async.dart';
import 'package:maybe_slow_future/maybe_slow_future.dart';
import 'package:test/test.dart';

void verifyOnSlowCallback({
  int expectedTimeToComplete,
  int actualTimeToComplete,
  bool expectSlowCallbackCalled,
}) {
  final expectedDurationToComplete = Duration(
    milliseconds: expectedTimeToComplete,
  );

  final actualDurationToComplete = Duration(
    milliseconds: actualTimeToComplete,
  );

  fakeAsync((fasync) {
    final longFuture = Future.delayed(
      actualDurationToComplete,
      () => 'dummyResult',
    );

    bool isSlowCallbackCalled = false;

    maybeSlowFuture(
      actualFuture: longFuture,
      threshold: expectedDurationToComplete,
      onSlowCallback: () => isSlowCallbackCalled = true,
    );

    fasync.elapse(const Duration(seconds: 1));

    expect(isSlowCallbackCalled, expectSlowCallbackCalled);
  });
}

void main() {
  group('maybeSlowFuture', () {
    test(
      'call onSlowCallback when a future fails to complete in the given threshold',
      () {
        verifyOnSlowCallback(
          expectedTimeToComplete: 250,
          actualTimeToComplete: 500,
          expectSlowCallbackCalled: true,
        );
      },
    );

    test(
      'do not call onSlowCallback when a future completes in the given threshold',
      () {
        verifyOnSlowCallback(
          expectedTimeToComplete: 250,
          actualTimeToComplete: 100,
          expectSlowCallbackCalled: false,
        );
      },
    );

    test(
      'assert actualFuture is required',
      () {
        expect(
          () => maybeSlowFuture(
            actualFuture: null,
            threshold: Duration(),
            onSlowCallback: () {},
          ),
          throwsA(isA<Error>()),
        );
      },
    );

    test(
      'assert threshold is required',
      () {
        expect(
          () => maybeSlowFuture(
            actualFuture: Future.value(0),
            threshold: null,
            onSlowCallback: () {},
          ),
          throwsA(isA<Error>()),
        );
      },
    );

    test(
      'assert onSlowCallback is required',
      () {
        expect(
          () => maybeSlowFuture(
            actualFuture: Future.value(0),
            threshold: Duration(),
            onSlowCallback: null,
          ),
          throwsA(isA<Error>()),
        );
      },
    );

    test('example of using extension method', () {
      fakeAsync((fasync) {
        final operation = Future.delayed(
          Duration(milliseconds: 1000),
          () => 'dummyValue',
        );

        bool isSlowOperation = false;
        // using extension variant
        operation.onSlow(Duration(milliseconds: 500), () {
          isSlowOperation = true;
        });

        operation.then((value) => expect(value, 'dummyValue'));

        fasync.elapse(Duration(milliseconds: 1000));

        expect(isSlowOperation, true);
      });
    });
  });
}


lib/maybe_slow_future.dart
library maybe_slow_future;

import 'dart:async';

/// Provide the extension method [onSlow] for a [Future]
extension OnSlowFuture<T> on Future<T> {
  /// Passing [threshold] and [onSlowCallback] to [maybeSlowFuture]
  Future<T> onSlow(Duration threshold, void Function() onSlowCallback) {
    return maybeSlowFuture(
      actualFuture: this,
      threshold: threshold,
      onSlowCallback: onSlowCallback,
    );
  }
}

/// Wrap a [Future] and return another [Future] that performs the [onSlowCallback]
/// when a certain time (specified by the [threshold] parameter) has passed.
///
/// It can be useful in certain scenario,
/// i.e. when we want to communicate to the user that an on-going operation is taking longer than it should be
Future<T> maybeSlowFuture<T>({
  Future<T> actualFuture,
  Duration threshold,
  void Function() onSlowCallback,
}) {
  assert(actualFuture != null);
  assert(threshold != null);
  assert(onSlowCallback != null);

  return actualFuture.timeout(threshold, onTimeout: () {
    onSlowCallback();
    return actualFuture;
  });
}


example/main.dart
/// Remember to import the package to be able to use the [Future.onSlow] extension method
import 'package:maybe_slow_future/maybe_slow_future.dart';

void main() {
  print('Fetching data...');

  const timeToComplete = Duration(seconds: 5);

  /// simulate a long running operation
  final fetchDataFuture = Future.delayed(timeToComplete, () {
    print('Got data after ${timeToComplete}: Hello World!');
  });

  /// use the extension method [Future.onSlow]
  fetchDataFuture.onSlow(Duration(seconds: 2), () {
    print('Be patient! It might take a while to complete...');
  });

  /// or use the [maybeSlowFuture] function
  maybeSlowFuture(
    actualFuture: fetchDataFuture,
    threshold: Duration(seconds: 3, milliseconds: 500),
    onSlowCallback: () {
      print('This is taking longer than expected...');
    },
  );

  /// output:
  /// Fetching data...
  /// Be patient! It might take awhile to complete...
  /// This is taking longer than expected...
  /// Got data after 0:00:05.000000: Hello World!
}
